File: app/src/main/java/network/ex02.java
package network;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.CountDownLatch;

public class ex02 {
    public static void main(String[] args) throws Exception {
        CountDownLatch ready = new CountDownLatch(1);
        int[] portHolder = new int[1];

        Thread server = new Thread(() -> {
            try (ServerSocket serverSocket = new ServerSocket(0)) {
                portHolder[0] = serverSocket.getLocalPort();
                ready.countDown();

                try (Socket client = serverSocket.accept();
                     BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
                     PrintWriter out = new PrintWriter(client.getOutputStream(), true)) {
                    String msg = in.readLine();
                    out.println("echo: " + msg);
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });

        server.start();
        ready.await();

        try (Socket socket = new Socket("127.0.0.1", portHolder[0]);
             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {
            out.println("hello tcp");
            System.out.println(in.readLine());
        }

        server.join();
    }
}

File: app-kotlin/src/main/kotlin/network/ex02.kt
package network

import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.PrintWriter
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.CountDownLatch

class ex02 {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val ready = CountDownLatch(1)
            val portHolder = IntArray(1)

            val server = Thread {
                ServerSocket(0).use { serverSocket ->
                    portHolder[0] = serverSocket.localPort
                    ready.countDown()

                    serverSocket.accept().use { client ->
                        val reader = BufferedReader(InputStreamReader(client.getInputStream()))
                        val writer = PrintWriter(client.getOutputStream(), true)
                        val msg = reader.readLine()
                        writer.println("echo: $msg")
                    }
                }
            }

            server.start()
            ready.await()

            Socket("127.0.0.1", portHolder[0]).use { socket ->
                val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
                val writer = PrintWriter(socket.getOutputStream(), true)
                writer.println("hello tcp")
                println(reader.readLine())
            }

            server.join()
        }
    }
}
